### ISSUE TITLE
Trimming tests should restore upfront instead of after the product build

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#54920]
See https://dev.azure.com/dnceng/public/_build/results?buildId=1211661&view=logs&jobId=9bd06b96-2bdb-52e1-9716-7806e46fa22b&j=9bd06b96-2bdb-52e1-9716-7806e46fa22b&t=a3f0b638-9032-505e-73db-bdcb5f500afb which was failing during restore but after the product build.

By moving to an up-front restore of the (generated) trimming test project, restore is bundled together with the rest of the product restore and NuGet can either fail early or succeed. Also restoring everything in one go makes the overall restore much more efficient. Failing late in the build makes both a poor local developer experience and wastes resources on CI.

cc @joperezr @safern @eerhardt

**Comment by safern** [comment]
How can we be able to restore generated trimming projects upfront? Do you plan on restoring just it's dependencies so that they are on the cache? Or were you thinking generating them first, then restore the product?

**Comment by ViktorHofer** [comment]
> Or were you thinking generating them first, then restore the product?

Yes. A target could be added to the InitialTargets of tests.proj that would make sure that test projects are incrementally generated and ProjectReferences point to them. That should do the job.
